pkg/controller/backupinfrastructure/backup_infrastructure_control.go

```diff
@@ -0,0 +1,338 @@
+	ReconcileBackupInfrastructure(backupInfrastructure *gardenv1beta1.BackupInfrastructure, key string) error
+}
+
+// NewDefaultControl returns a new instance of the default implementation ControlInterface that
+// implements the documented semantics for BackupInfrastructures. You should use an instance returned from NewDefaultControl() for any
+// scenario other than testing.
+func NewDefaultControl(k8sGardenClient kubernetes.Client, k8sGardenInformers gardeninformers.Interface, secrets map[string]*corev1.Secret, imageVector imagevector.ImageVector, identity *gardenv1beta1.Gardener, config *componentconfig.ControllerManagerConfiguration, recorder record.EventRecorder, updater UpdaterInterface) ControlInterface {
+	return &defaultControl{k8sGardenClient, k8sGardenInformers, secrets, imageVector, identity, config, recorder, updater}
+}
+
+type defaultControl struct {
+	k8sGardenClient    kubernetes.Client
+	k8sGardenInformers gardeninformers.Interface
+	secrets            map[string]*corev1.Secret
+	imageVector        imagevector.ImageVector
+	identity           *gardenv1beta1.Gardener
+	config             *componentconfig.ControllerManagerConfiguration
+	recorder           record.EventRecorder
+	updater            UpdaterInterface
+}
+
+func (c *defaultControl) ReconcileBackupInfrastructure(obj *gardenv1beta1.BackupInfrastructure, key string) error {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return err
+	}
+
+	var (
+		backupInfrastructure       = obj.DeepCopy()
+		backupInfrastructureLogger = logger.NewFieldLogger(logger.Logger, "backupinfrastructure", fmt.Sprintf("%s/%s", backupInfrastructure.Namespace, backupInfrastructure.Name))
+		lastOperation              = backupInfrastructure.Status.LastOperation
+		operationType              = controllerutils.ComputeOperationType(lastOperation)
+	)
+
+	logger.Logger.Infof("[BACKUPINFRASTRUCTURE RECONCILE] %s", key)
+	backupInfrastructureJSON, _ := json.Marshal(backupInfrastructure)
+	backupInfrastructureLogger.Debugf(string(backupInfrastructureJSON))
+
+	op, err := operation.NewWithBackupInfrastructure(backupInfrastructure, backupInfrastructureLogger, c.k8sGardenClient, c.k8sGardenInformers, c.identity, c.secrets, c.imageVector, c.config)
+	if err != nil {
+		backupInfrastructureLogger.Errorf("Could not initialize a new operation: %s", err.Error())
+		return err
+	}
+
+	// The deletionTimestamp labels a BackupInfrastructure as intended to get deleted. Before deletion,
+	// it has to be ensured that no infrastructure resources are depending on the BackupInfrastructure anymore.
+	// When this happens the controller will remove the finalizer from the BackupInfrastructure so that it can be garbage collected.
+	if backupInfrastructure.DeletionTimestamp != nil {
+		gracePeriod := time.Hour * 24 * time.Duration(*c.config.Controllers.BackupInfrastructure.DeletionGracePeriodDays)
+		if time.Now().Sub(backupInfrastructure.DeletionTimestamp.Time) > gracePeriod {
```

Comment:
Yes we need have somewhere indication about actual deletion time. I will keep LastOperationState : `Pending` (Sorry i couldn't think of better)LastOperationType `Delete` and the description: "Actual deletion will happen at (deletionTimestamp+gracePeriod) " only at the successful reconciliation in grace period. While reconciliation the state will be still `progressing` and type will be `delete`.

Reply:
hm, shouldn't it be `type=delete` and `state=pending`?

Reply:
No, once deletionTimestamp set,
Within deletion grace period: Type will be `delete`, while reconciling state should be `processing`. And after successful reconciliation it should be `pending`. If we keep both `pending` then it would be confusing, since we won't be able to say reconciliation is happening or it's done.

Reply:
How about `suspended` instead of `pending`?

Reply:
I don't get it. `state=processing` indicates that an operation is currently executed on the resource. We shouldn't keep `state=processing` for the deletion grace period time window (up to days or even months) as simply nothing is actually processing. I'll prefer, as said, `type=delete`, `state=pending` and `description=Deletion of bucket scheduled for <datetime>`.

Reply:
both is fine, doesn't matter